src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexInterpreter.cs

```diff
@@ -223,6 +224,20 @@ private bool MatchString(string str)
             {
                 if (runtextend - runtextpos < c)
                 {
+                    // If MatchString was called after a greedy op such as a .*, we would have zipped runtextpos to the end without really examining any characters. Reset to maxBacktrackPos here as an optimization
+                    if (!_caseInsensitive && _maxBacktrackPosition != -1 && runtextpos > _maxBacktrackPosition)
```

Comment:
I'm probably just missing it, but what is it that prevents this from kicking in if you have something between the loop and the string, e.g. `.*[abc]string`?

Reply:
Ok, I finally have enough time to spend on this without context switching. The answer to your question is the `(runtextend - runtextpos < c)` part. I tested `.*[abc]string` against the text `fooastring`. `_code` for `.*[abc]string` looks like this:

```
0: [^\x0a]

000000 *Lazybranch       addr = 14
000002 *Setmark          
000003 *Notoneloop       '\x0a', rep = inf
000006  UpdateBumpalong  
000007  Set              [a-c]
000009  Multi            "string"
000011 *Capturemark      index = 0
000014  Stop             
```

The Notoneloop will first take us past the `g`. Then we backtrack 1 by 1 (another potential optimization probably exists here) till the `a`. `a` matches the set [a-c], so we bump up by 1 and go to `Multi`. In `MatchString`, `runtextend = 10, runtextpos = 4`, so we never enter the `if` block.